addons/project/models/project_task.py

```diff
@@ -1147,6 +1147,23 @@ def write(self, vals):
         if "personal_stage_type_id" in vals and not vals['personal_stage_type_id']:
             del vals['personal_stage_type_id']
 
+        # sends an email to the 'Task Creation' subtype subscribers
+        # When project_id is changed
+        if vals.get('project_id'):
+            project = self.env['project.project'].browse(vals.get('project_id'))
+            notification_subtype_id = self.env['ir.model.data']._xmlid_to_res_id('project.mt_project_task_new')
+            partner_ids = project.message_follower_ids.filtered(lambda follower: notification_subtype_id in follower.subtype_ids.ids).partner_id.ids
+            if partner_ids:
+                link_per_project_id = {p.id: p._get_html_link(title=p.display_name) for p in self.project_id}
+                for task in self:
+                    project_link = link_per_project_id[task.project_id.id]
+                    task.message_notify(
+                        body=_('Task Transferred from Project %s', project_link),
+                        partner_ids=partner_ids,
+                        email_layout_xmlid='mail.mail_notification_layout',
+                        record_name=task.display_name,
+                    )
```

Comment:
In fact, since we send the notification before updating the record, the write could fail in the parent classes and so the notification will be wrong since the task does not move. It would be better to do it after having called `super`.

Also, if the task is in fact a private one (= a To-Do) then the line with `project_link = link_per_project_id[task.project_id.id]` will raise a traceback since `False` will not be found in `link_per_project`.

I discussed with Alexandra about that and we conclude to display another message for the private task converted into a basic task. We will send a notification with the following message: `Task Converted from To-Do`.

Here is the diff applying all my remarks:

```diff
diff --git a/addons/project/models/project_task.py b/addons/project/models/project_task.py
index 579aa108ff70..b2fe934412e9 100644
--- a/addons/project/models/project_task.py
+++ b/addons/project/models/project_task.py
@@ -1003,6 +1003,8 @@ class Task(models.Model):
         if len(self) == 1:
             handle_history_divergence(self, 'description', vals)
         portal_can_write = False
+        project_link_per_task_id = {}
+        partner_ids = []
         if self.env.user.has_group('base.group_portal') and not self.env.su:
             # Check if all fields in vals are in SELF_WRITABLE_FIELDS
             self._ensure_fields_are_accessible(vals.keys(), operation='write', check_group_user=False)
@@ -1139,15 +1141,13 @@ class Task(models.Model):
             notification_subtype_id = self.env['ir.model.data']._xmlid_to_res_id('project.mt_project_task_new')
             partner_ids = project.message_follower_ids.filtered(lambda follower: notification_subtype_id in follower.subtype_ids.ids).partner_id.ids
             if partner_ids:
-                link_per_project_id = {p.id: p._get_html_link(title=p.display_name) for p in self.project_id}
+                link_per_project_id = {}
                 for task in self:
-                    project_link = link_per_project_id[task.project_id.id]
-                    task.message_notify(
-                        body=_('Task Transferred from Project %s', project_link),
-                        partner_ids=partner_ids,
-                        email_layout_xmlid='mail.mail_notification_layout',
-                        record_name=task.display_name,
-                    )
+                    if task.project_id:
+                        project_link = link_per_project_id.get(task.project_id.id)
+                        if not project_link:
+                            project_link = link_per_project_id[task.project_id.id] = task.project_id._get_html_link(title=task.project_id.display_name)
+                        project_link_per_task_id[task.id] = project_link

         result = super().write(vals)

@@ -1177,6 +1177,20 @@ class Task(models.Model):
             self.filtered(lambda t: t.state != '04_waiting_normal').state = '01_in_progress'

         self._task_message_auto_subscribe_notify({task: task.user_ids - old_user_ids[task] - self.env.user for task in self})
+
+        if partner_ids:
+            for task in self:
+                project_link = project_link_per_task_id.get(task.id)
+                if project_link:
+                    body = _('Task Transferred from Project %s', project_link)
+                else:
+                    body = _('Task Converted from To-Do')
+                task.message_notify(
+                    body=body,
+                    partner_ids=partner_ids,
+                    email_layout_xmlid='mail.mail_notification_layout',
+                    record_name=task.display_name,
+                )
         return result

     def unlink(self):
```

cc @MissingNoShiny